README.md
# gradient_like_css

The `gradient_like_css` package for Flutter allows you to experience CSS-like gradients in your Flutter app.

<br>

## Installing

### 1. Depend on it

Add this to your package's `pubspec.yaml` file:

```yaml
dependencies:
  gradient_like_css: ^1.0.0
```

### 2. Install it

You can install packages from the command line:

with `Flutter`:

```
$ flutter pub get
```

### 3. Import it

Now in your `Dart` code, you can use:

```dart
import 'package:gradient_like_css/gradient_like_css.dart';
```

<br>

## Usage

To import `CssLike`:

```dart
import 'package:gradient_like_css/gradient_like_css.dart';
```

To use `CssLike` with the `BoxDecoration`:

```dart
BoxDecoration(
  gradient: linearGradient(-225, ['#69EACB', '#EACCF8 48%', "#6654F1"]),
);
```

<br> 

## Example

### `CssLike.linearGradient()`

#### 🎨 Gradient at the default angle

<img src="https://raw.githubusercontent.com/azukisiromochi/gradient_like_css/main/display/CssLike_linearGradient__example1.png" align="right" height="300px">

with `CSS`:

```css
background: linear-gradient(#e66465, #9198e5);
```

with `Flutter`:

```dart
Container(
  height: 300,
  width: 300,
  decoration: BoxDecoration(
    gradient: linearGradient(null, ['#e66465', '#9198e5']),
  ),
);
```

>  _\*Note: If the first argument is `null`, a 180 degree angular gradient is created._

<br>

#### 🎨 Gradient at a 45-degree angle

<img src="https://raw.githubusercontent.com/azukisiromochi/gradient_like_css/main/display/CssLike_linearGradient__example2.png" align="right" height="300px">

with `CSS`:

```css
background: linear-gradient(45deg, red, blue);
```

with `Flutter`:

```dart
Container(
  height: 300,
  width: 300,
  decoration: BoxDecoration(
    gradient: linearGradient(45, ['red', 'blue']),
  ),
);
```

>  _\*Note: The `color` argument  can use [X11/CSS3 color](https://en.wikipedia.org/wiki/Web_colors#X11_color_names) names._

<br>

#### 🎨 Gradient that starts at 60% of the gradient line

<img src="https://raw.githubusercontent.com/azukisiromochi/gradient_like_css/main/display/CssLike_linearGradient__example3.png" align="right" height="300px">

with `CSS`:

```css
background: linear-gradient(135deg, orange, orange 60%, cyan);
```

with `Flutter`:

```dart
Container(
  height: 300,
  width: 300,
  decoration: BoxDecoration(
    gradient: linearGradient(135, ['orange', 'orange 60%', 'cyan']),
  ),
);
```

>  _\*Note: You can add a color-stop points using the `stop` argument. It can be used with % such as `'orange 60%'`._

<br>

#### 🎨 Gradient with multi-position color stops

<img src="https://raw.githubusercontent.com/azukisiromochi/gradient_like_css/main/display/CssLike_linearGradient__example4.png" align="right" height="300px">

with `CSS`:

```css
background: linear-gradient(to right,
     red 20%, orange 20% 40%, yellow 40% 60%, green 60% 80%, blue 80%);
```

with `Flutter`:

```dart
Container(
  height: 300,
  width: 300,
  decoration: BoxDecoration(
    gradient: linearGradient(Alignment.centerRight, 
      ['red 20%', 'orange 20% 40%', 'yellow 40% 60%', 'green 60% 80%', 'blue 80%']),
  ),
);
```

>  _\*Note: The first argument can be `Alignment` as well as angle._

<br>

### `WebColors` / `X11Colors`

To use `X11Colors`:

```dart
Container(
    color: X11Colors.MediumSpringGreen.color,
);
```

To use `WebColors`  by [X11/CSS3 color](https://en.wikipedia.org/wiki/Web_colors#X11_color_names) names:

```dart
Container(
    // Can be used in lowercase too
    color: WebColors.of('MediumSpringGreen').color,
);
```

<br>

## Features

- `radialGradient()`
- `conicGradient()` ( `sweepGradient`  )

<br>

## Bugs or Requests

If you encounter any problems feel free to open an [issue](https://github.com/azukisiromochi/gradient_like_css/issues/new?template=bug_report.md). If you feel the library is missing a feature, please raise a [ticket](https://github.com/azukisiromochi/gradient_like_css/issues/new?template=feature_request.md) on GitHub and I'll look into it. Pull request are also welcome.

CHANGELOG.md
## [1.0.1] - 2022-05-25

- 📚 Update the README to change the version rule

## [1.0.0-nullsafety.0] - 2022-05-25

- ✨ Add null safety support

## [0.5.1] - 2021-01-26

- 🐛 Fixed color stop point bug

## [0.5.0] - 2021-01-26

- 🚨  Unit test finished

## [0.1.0] - 2021-01-22

- 📚 Preparing for release

## [0.0.5] - 2021-01-22

- 📚 Add changelog
- 📚 Update README with import instructions and example
- 📚 README image path fixes

## [0.0.4] - 2021-01-22

- 📚 Add license

## [0.0.3] - 2021-01-22

- ♻ Change property name, etc

## [0.0.2] - 2021-01-21

- ✨ Add to `<side-or-corner>` parameter
  - In CSS, `linear-gradient` has a `<side-or-corner>` parameter. Now available in Flutter.

## [0.0.1] - 2021-01-21

* 🎉 Make Flutter plugin project
  * [WIP] Make the `CssLike` class, which is the main feature of this plugin.
  * Make the `WebColor` class that uses [X11/CSS3 colors](https://en.wikipedia.org/wiki/Web_colors#X11_color_names).



test/src/css_like_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import '../../lib/src/css_like.dart';

void main() {
  group('angleOrEndAlignment', () {
    test('Angle is int #1', () {
      const expected = LinearGradient(
        begin: Alignment.bottomCenter,
        end: Alignment.topCenter,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals
      int intValue = 0;
      final actual = linearGradient(intValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is int #2', () {
      const expected = LinearGradient(
        begin: Alignment.centerLeft,
        end: Alignment.centerRight,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals
      int intValue = 90;
      final actual = linearGradient(intValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is int #3', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals
      int intValue = 180;
      final actual = linearGradient(intValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is int #4', () {
      const expected = LinearGradient(
        begin: Alignment.centerRight,
        end: Alignment.centerLeft,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals
      int intValue = 270;
      final actual = linearGradient(intValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is int #5', () {
      const expected = LinearGradient(
        begin: Alignment.bottomCenter,
        end: Alignment.topCenter,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals
      int intValue = 360;
      final actual = linearGradient(intValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is double', () {
      const expected = LinearGradient(
        begin: Alignment.bottomLeft,
        end: Alignment.topRight,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals, prefer_int_literals
      double doubleValue = 45.0;
      final actual = linearGradient(doubleValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is angle over 360 degrees', () {
      const expected = LinearGradient(
        begin: Alignment.bottomLeft,
        end: Alignment.topRight,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      const angleOver360Degrees = 405.0;
      final actual =
          linearGradient(angleOver360Degrees, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is negative number #1', () {
      const expected = LinearGradient(
        begin: Alignment.bottomLeft,
        end: Alignment.topRight,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals, prefer_int_literals
      double doubleValue = -315.0;
      final actual = linearGradient(doubleValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is negative number #2', () {
      const expected = LinearGradient(
        begin: Alignment.centerRight,
        end: Alignment.centerLeft,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals, prefer_int_literals
      double doubleValue = -90.0;
      final actual = linearGradient(doubleValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is negative number #3', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals, prefer_int_literals
      double doubleValue = -180.0;
      final actual = linearGradient(doubleValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is negative number #4', () {
      const expected = LinearGradient(
        begin: Alignment.centerLeft,
        end: Alignment.centerRight,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals, prefer_int_literals
      double doubleValue = -270.0;
      final actual = linearGradient(doubleValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('Angle is negative number #5', () {
      const expected = LinearGradient(
        begin: Alignment.bottomCenter,
        end: Alignment.topCenter,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      // ignore: omit_local_variable_types, prefer_final_locals, prefer_int_literals
      double doubleValue = -360.0;
      final actual = linearGradient(doubleValue, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('EndAlignment is normal value', () {
      const expected = LinearGradient(
        begin: Alignment.centerRight,
        end: Alignment.centerLeft,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      final actual =
          linearGradient(Alignment.centerLeft, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('angleOrEndAlignment is null', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      final actual = linearGradient(null, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('angleOrEndAlignment is Object', () {
      expect(() => linearGradient(Object(), ['#e66465', '#9198e5']),
          throwsFormatException);
    });
  });

  group('colorStopList', () {
    test('colorStopList has one argument', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      final actual = linearGradient(null, ['#e66465', '#9198e5']);

      expect(actual, expected);
    });
    test('colorStopList has two arguments', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.2,
          1.0,
        ],
      );

      final actual = linearGradient(null, ['#e66465 20%', '#9198e5']);

      expect(actual, expected);
    });
    test('colorStopList has three arguments', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffe66465),
          Color(0xffe66465),
          Color(0xff9198e5),
        ],
        stops: [
          0.2,
          0.5,
          1.0,
        ],
      );

      final actual = linearGradient(null, ['#e66465 20% 50%', '#9198e5']);

      expect(actual, expected);
    });
    test('colorStopList has a 3-digit Hex code', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffee6655),
          Color(0xffff9911),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      final actual = linearGradient(null, ['#e65', '#f91']);

      expect(actual, expected);
    });
    test('colorStopList use a web color names', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffff1493),
          Color(0xffdcdcdc),
        ],
        stops: [
          0.0,
          1.0,
        ],
      );

      final actual = linearGradient(null, ['DeepPink', 'Gainsboro']);

      expect(actual, expected);
    });
    test('completion of stop', () {
      const expected = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [
          Color(0xffffa500),
          Color(0xffffff00),
          Color(0xff98fb98),
          Color(0xff00ffff),
        ],
        stops: [
          0.2,
          0.4,
          0.6,
          0.8,
        ],
      );

      final actual = linearGradient(
          null, ['orange 20%', 'yellow', 'paleGreen', 'cyan 80%']);

      expect(actual, expected);
    });
    test('colorStopList has an invalid value', () {
      expect(() => linearGradient(null, ['#e66465 20', '#9198e5']),
          throwsFormatException);
      expect(() => linearGradient(null, ['#e66465 A%', '#9198e5']),
          throwsFormatException);
    });
    test('colorStopList is null', () {
      expect(() => linearGradient(null, null), throwsFormatException);
    });
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import '../lib/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/gradient_like_css_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:gradient_like_css/gradient_like_css.dart';

void main() {
  const channel = MethodChannel('gradient_like_css');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('CssLike#linearGradient()', () {
    expect(
        CssLike.linearGradient(
            90, ['#FFE5C5 17.61%', '#D2B38B 50.22%', '#F1DDC3 83.52%']),
        const LinearGradient(
          begin: Alignment(0, -1),
          end: Alignment(0, 1),
          colors: [
            const Color(0xffFFE5C5),
            const Color(0xffD2B38B),
            const Color(0xffF1DDC3),
          ],
          stops: const [
            0.1761,
            0.5022,
            0.8352,
          ],
        ));
  });
}


lib/src/css_like.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';
import 'package:gradient_like_css/src/web_colors.dart';
import 'package:tuple/tuple.dart';

///
/// Create linear gradients with CSS-like coding.
/// For example, the following code.
///
/// ```dart
/// BoxDecoration(
///   gradient: linearGradient(-225, ['#69EACB', '#EACCF8 48%', "#6654F1"]),
/// );
///```
///
/// The [angleOrEndAlignment] argument is valid only for the [double] or
/// [Alignment] type.
///
/// The [colorStopList] argument is a space-separated list of String
/// containing colors and stops.
/// The [colorStopList] argument must not be null. Colors allow web color
/// names or color codes that start with "#". Stops allow percentage strings
/// like "12.34%".
///
LinearGradient linearGradient(
  Object? angleOrEndAlignment,
  List<String>? colorStopList,
) {
  final endAlignment = _getEndAlignment(angleOrEndAlignment);
  final colorsAndStops = _getColorsAndStops(colorStopList);

  return LinearGradient(
    begin: -endAlignment,
    end: endAlignment,
    colors: colorsAndStops.item1,
    stops: colorsAndStops.item2,
  );
}

Alignment _getEndAlignment(Object? angleOrEndAlignment) {
  if (angleOrEndAlignment == null) {
    return Alignment.bottomCenter;
  } else if (angleOrEndAlignment is num) {
    final angle = angleOrEndAlignment.toDouble();
    return _degreesToAlignment(angle -90);
  } else if (angleOrEndAlignment is Alignment) {
    return angleOrEndAlignment;
  } else {
    throw const FormatException(
      // ignore: lines_longer_than_80_chars
      'The "angleOrEndAlignment" argument is valid only for the "double" or "Alignment" type.',
    );
  }
}

Tuple2<List<Color>, List<double>> _getColorsAndStops(
  List<String>? colorStopList,
) {
  final colors = <Color>[];
  final stops = <double>[];

  if ((colorStopList ?? const <String>[]).isEmpty) {
    throw const FormatException(
      // ignore: lines_longer_than_80_chars
      'The "colorStopList" argument can be set up to three, separated by spaces, such as "yellow 40% 60%".',
    );
  }

  for (final param in colorStopList!) {
    var colorCode = '';
    var percentage1 = '';
    var percentage2 = '';

    final splitParam = param.split(' ');
    if (splitParam.isEmpty || splitParam.length > 3) {
      throw const FormatException(
        // ignore: lines_longer_than_80_chars
        'The "colorStopList" argument can be set up to three, separated by spaces, such as "yellow 40% 60%".',
      );
    }
    if (splitParam.isNotEmpty) {
      colorCode = splitParam[0];
    }
    if (splitParam.length > 1) {
      percentage1 = splitParam[1];
    }
    if (splitParam.length > 2) {
      percentage2 = splitParam[2];
    }

    final color = _codeToColor(colorCode);
    final stop1 = _percentageStringToStop(percentage1);
    if (percentage2.isEmpty) {
      colors.add(color);
      stops.add(stop1);
    } else {
      colors
        ..add(color)
        ..add(color);
      stops
        ..add(stop1)
        ..add(_percentageStringToStop(percentage2));
    }
  }

  if (stops.first.isNaN) {
    stops.first = 0.0;
  }
  if (stops.last.isNaN) {
    stops.last = 1.0;
  }
  stops.asMap().forEach((index, stop) {
    if (stop.isNaN) {
      final start = index;
      var end = start;
      while (stops[end + 1].isNaN) {
        end++;
      }
      final previousStop = stops[start - 1];
      final nextStop = stops[end + 1];
      final range = end - index + 1;
      final separation = (nextStop - previousStop) / (range + 1);

      for (var i = 0; i < range; i++) {
        stops[index + i] = double.parse(
          (previousStop + separation * (i + 1)).toStringAsPrecision(8),
        );
      }
    }
  });
  return Tuple2<List<Color>, List<double>>(colors, stops);
}

double _percentageStringToStop(String percentageString) {
  if (percentageString.isEmpty) {
    return double.nan;
  }
  if (!percentageString.endsWith('%')) {
    throw const FormatException(
      'Bad stop format (Allow percentage strings like "12.34%").',
    );
  }

  try {
    final stop = double.parse(percentageString.replaceAll('%', '')) / 100;
//      assert(0.0 <= stop && stop <= 1.0);
    return stop;
  } on Exception {
    throw const FormatException(
      'Bad stop format (Allow percentage strings like "12.34%").',
    );
  }
}

Color _codeToColor(String code) {
  final webColor = WebColors.of(code);
  if (webColor != X11Colors.None) {
    return webColor.color;
  }
  return Color(_makeHexCode(code));
}

int _makeHexCode(String code) {
  final hexColorCodeExp = RegExp(r'^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$');
  if (hexColorCodeExp.allMatches(code).length != 1) {
    throw const FormatException(
      // ignore: lines_longer_than_80_chars
      'Bad color code format (Allow web color name or color code that start with "#").',
    );
  }

  String hexCode;
  if (code.length == 4) {
    final r = code[1];
    final g = code[2];
    final b = code[3];
    hexCode = '0xFF$r$r$g$g$b$b';
  } else {
    hexCode = code.replaceFirst('#', '0xFF');
  }

  return int.parse(hexCode);
}

Alignment _degreesToAlignment(double degrees) {
  final verticalOrHorizontal = _getVerticalOrHorizontal(degrees);
  if (verticalOrHorizontal != null) {
    return verticalOrHorizontal;
  }

  final x = _x(degrees);
  final y = _y(degrees);
  final xAbs = x.abs();
  final yAbs = y.abs();

  if ((0.0 < xAbs && xAbs < 1.0) || (0.0 < yAbs && yAbs < 1.0)) {
    final magnification = (1 / xAbs) < (1 / yAbs) ? (1 / xAbs) : (1 / yAbs);
    return Alignment(x, y) * magnification;
  } else {
    return Alignment(x, y);
  }
}

Alignment? _getVerticalOrHorizontal(double degrees) {
  var modDeg = degrees % 360;
  if (degrees < 0.0 && modDeg != 0.0) {
    modDeg -= 360;
  }

  if (modDeg == 0.0 || modDeg == -0.0) {
    return Alignment.centerRight;
  }
  if (modDeg == 90.0 || modDeg == -270.0) {
    return Alignment.bottomCenter;
  }
  if (modDeg == 180.0 || modDeg == -180.0) {
    return Alignment.centerLeft;
  }
  if (modDeg == 270.0 || modDeg == -90.0) {
    return Alignment.topCenter;
  }
  return null;
}

double _x(double degrees) {
  final radians = degrees / 180.0 * math.pi;
  return double.parse(math.cos(radians).toStringAsPrecision(8));
}

double _y(double degrees) {
  final radians = degrees / 180.0 * math.pi;
  return double.parse(math.sin(radians).toStringAsPrecision(8));
}


android/src/main/java/com/azukisiromochi/gradient_like_css/GradientLikeCssPlugin.java
package com.azukisiromochi.gradient_like_css;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** GradientLikeCssPlugin */
public class GradientLikeCssPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "gradient_like_css");
    channel.setMethodCallHandler(this);
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "gradient_like_css");
    channel.setMethodCallHandler(new GradientLikeCssPlugin());
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:gradient_like_css/gradient_like_css.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('gradient_like_css'),
        ),
        body: Center(
          child: Container(
            height: 300,
            width: 300,
            decoration: BoxDecoration(
              gradient: linearGradient(null, ['#e66465', '#9198e5']),
              /// Gradient at a 45-degree angle
              // gradient: linearGradient(45, ['red', 'blue']),
              /// Gradient that starts at 60% of the gradient line
              // gradient: linearGradient(135, ['orange', 'orange 60%', 'cyan']),
              /// Gradient with multi-position color stops
              // gradient: linearGradient(
              //   Alignment.centerRight,
              //   [
              //     'red 20%',
              //     'orange 20% 40%',
              //     'yellow 40% 60%',
              //     'green 60% 80%',
              //     'blue 80%',
              //   ],
              // ),
            ),
          ),
        ),
      ),
    );
  }
}
